File: 143-reorder-list/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head->next;
        // getting middle element
        while(fast!=NULL && fast->next!=NULL){
            slow=slow->next;
            fast = fast->next->next;
        }
        ListNode* mid = slow->next;
        slow->next = NULL;// so that connection cut at half
        //reversing from middle element
        ListNode* prev = NULL;
        ListNode* current = mid;
        while(current!=NULL){
            ListNode* Next = current->next;
            current->next = prev;
            prev = current;
            current = Next;

        }
       ListNode* node2 = prev;// gives the reverse of LL starting with end element
       ListNode* node1 = head;

       while( node1!=NULL && node2!=NULL){
        ListNode* temp1 = node1->next;
        ListNode* temp2 = node2->next;
        node1->next = node2;
        node1 = temp1;
        node2->next = node1;
        node2 = temp2;
       }
           
    }
};

File: 143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        
        ListNode fast = head.next;
        ListNode slow = head;
        ListNode prev = null;
        while(fast!=null && fast.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode newH = slow.next;
        slow.next = null;
        newH = reverse(newH);
        slow = head;
        while(newH !=null)
        {
            ListNode n1 = slow.next;
            ListNode n2 = newH.next;
            slow.next = newH;
            newH.next = n1 ;
            slow = n1;
            newH = n2;
        }
    }
    private ListNode reverse(ListNode head){
        ListNode prev = null;
        ListNode current = head;
        while(current!=null){
            ListNode n = current.next;
            current.next = prev;
            prev = current;
            current = n;
        }
        return prev;
    }
}